scripts/stackstorm.js

```diff
@@ -126,16 +124,30 @@ module.exports = function(robot) {
     };
   }
 
-  var api = st2client(opts);
+  var api_client = st2client(opts);
 
   if (env.ST2_API_KEY) {
-    api.setKey({ key: env.ST2_API_KEY });
+    api_client.setKey({ key: env.ST2_API_KEY });
   } else if (env.ST2_AUTH_TOKEN) {
-    api.setToken({ token: env.ST2_AUTH_TOKEN });
+    api_client.setToken({ token: env.ST2_AUTH_TOKEN });
+  }
+
+  function logErrorAndExit(err, res) {
+    if (err) {
+      robot.logger.error(err.stack);
+    }
+    if (res) {
```

Comment:
I found that second param `res` is only used in `robot.error` Hubot handler for uncaught exceptions.

Does it make sense to move entire `logErrorAndExit` code into `robot.error` handler itself?
And from the code where we want to exit just `throw err` as it was before. Looks better in the context of https://stackoverflow.com/a/40321400/4533625 (case 3) discussion.
It will be caught by `robot.error` then.

By throwing when we want to end an app (which will then call `robot.error`) I think allows any other bubbled up tasks to actually finish in a more safe way. It'll also stick with the previous code logic primitives. 

@blag What do you think?

Reply:
I considered this and ultimately decided against it. We use `logErrorAndExit` in one more place than just the Hubot error callback, and I suspect that we will expand the use of it in further commits.

Reply:
@jinpingh What do you think?

Reply:
Did run test case posted on https://github.com/StackStorm/st2chatops/issues/124#issuecomment-494075130. With @blag yesterday fix,  for invalid `ST2_API_URL` , `ST2_STREAM_URL` and `ST2_AUTH_TOKEN` , `st2chatops` logs out error message and keep waiting.   Is this the result we wanted?

Reply:
I'm actually talking here about reverting initial design where we previously `throw`'ed error as a way to stop hubot execution.
In our case `robot.error` handler will automatically catch those exceptions as unhandled and then call `stop`.

Here is the diff:


```diff
- robot.error(logErrorAndExit);

- function logErrorAndExit(err, res) {
+ robot.error(function(err, res) {
    if (err) {
      robot.logger.error(err.stack);
    }
    if (res) {
      res.send(JSON.stringify({
        "status": "failed",
        "msg": "An error occurred trying to post the message:\n" + err
      }));
    }

    stop({shutdown: true});
- };
+ });

- logErrorAndExit('Failed to authenticate: ' + err.message);
+ throw 'Failed to authenticate: ' + err.message;
```

By technical reasons I think that raising `robot.error` handler which then calls `stop()` via `throw` instead of `logErrorAndExit` function would align better with the existing shutdown mechanics.

Can add here cases when users want to re-use `hubot-stackstorm` module and so adding ability to catch errors and re-define `robot.error` on their side would make sense.

Reply:
> Add here cases when users want to re-use hubot-stackstorm module and so adding ability to catch errors and re-define robot.error on their side would make sense.

That isn't quite how the new code works.

Calling `robot.error` simply adds a callback function to the callback chain for the `error` event. See [this code](https://github.com/hubotio/hubot/blob/master/src/robot.js#L194) in Hubot for more.

If our users want to change or remove any of the error callbacks, they can still do so, they just have to manipulate `robot.errorHandlers` themselves. It's not as simple as assigning a function to `robot.onerror` anymore.

Reply:
Making the hubot error handler an unnamed lambda function, and then simply re-`throw()`ing exceptions is fine.